File: Tests/ConsoleRunnerTest.php
<?php

require_once dirname(__FILE__).'/../Yarrow/Autoload.php';

class ConsoleRunnerStub extends ConsoleRunner {
	
	function runInputProcess() {
		return;
	}
	
	function runOutputProcess() {
		return;
	}
	
}

class ConsoleRunnerTest extends PHPUnit_Framework_TestCase {
	
	function setUp() {
		ob_start();
	}
	
	function tearDown() {
		ob_end_clean();
	}
	
	function getVersionHeader() {
		return ConsoleRunnerStub::APPNAME . ' ' . ConsoleRunnerStub::VERSION;
	}
	
	function assertExitSuccess($actual) {
		$this->assertEquals(0, $actual);
	}
	
	function assertExitFailure($actual) {
		$this->assertEquals(1, $actual);
	}
	
	function testVersionMessage() {
		$app = new ConsoleRunnerStub(array('yarrow', '-v'));
		$this->assertExitSuccess($app->runApplication());
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
	}

	function testVersionMessageLong() {
		$app = new ConsoleRunnerStub(array('yarrow', '--version'));
		$this->assertExitSuccess($app->runApplication());
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
	}	
	
	function testHelpMessage() {
		$app = new ConsoleRunnerStub(array('yarrow', '-h'));
		$this->assertExitSuccess($app->runApplication());
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$this->assertContains('Path to the generated documentation', $output);
		$this->assertContains('Use -o or --option for boolean switches', $output);
	}
	
	function testHelpMessageLong() {
		$app = new ConsoleRunnerStub(array('yarrow', '--help'));
		$this->assertExitSuccess($app->runApplication());
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$this->assertContains('Path to the generated documentation', $output);
		$this->assertContains('Use -o or --option for boolean switches', $output);
	}
	
	function testInvalidOptionMessage() {
		$app = new ConsoleRunnerStub(array('yarrow', '-j'));
		$this->assertExitFailure($app->runApplication());
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$this->assertContains('Unrecognized option: -j', $output);
	}
	
	function testInvalidOptionMessageLong() {
		$app = new ConsoleRunnerStub(array('yarrow', '--invalid'));
		$this->assertExitFailure($app->runApplication());		
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$this->assertContains('Unrecognized option: --invalid', $output);
	}
	
	function testConfigMessageIfFileNotExists() {
		$app = new ConsoleRunnerStub(array('yarrow', '-c:missing.conf'));
		$this->assertExitFailure($app->runApplication());		
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$this->assertContains('File not found: missing.conf', $output);
	}
	
	function testConfigLoadsIfFileExists() {
		$testconfig = dirname(__FILE__).'/Config/.yarrowdoc';
		$app = new ConsoleRunnerStub(array('yarrow', '--config='.$testconfig));
		$this->assertExitSuccess($app->runApplication());		
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$this->assertNotContains('File not found', $output);
	}
	
	function testInputTargetsUpdatedFromSingleArgument() {
		$app = new ConsoleRunnerStub(array('yarrow', 'MyDocs'));
		$this->assertExitSuccess($app->runApplication());
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$config = Configuration::instance();
		$this->assertContains('MyDocs', $config->inputTargets);
	}

	function testInputTargetsUpdatedFromArgument() {
		$classpath = dirname(__FILE__).'/../Yarrow';
		$app = new ConsoleRunnerStub(array('yarrow', $classpath, 'MyDocs'));
		$this->assertExitSuccess($app->runApplication());		
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$config = Configuration::instance();
		$this->assertContains($classpath, $config->inputTargets);
		$this->assertEquals('MyDocs', $config->outputTarget);		
	}
	
	function testInputTargetsUpdatedFromArguments() {
		$classpath1 = dirname(__FILE__).'/../Yarrow';
		$classpath2 = dirname(__FILE__).'/Samples';
		$app = new ConsoleRunnerStub(array('yarrow', $classpath1, $classpath2, 'MyDocs'));
		$this->assertExitSuccess($app->runApplication());		
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$config = Configuration::instance();
		$this->assertEquals(2, count($config->inputTargets));
		$this->assertContains($classpath1, $config->inputTargets);
		$this->assertContains($classpath2, $config->inputTargets);
		$this->assertEquals('MyDocs', $config->outputTarget);		
	}
	
	function testInputTargetErrorFromBadArguments() {
		$classpath = dirname(__FILE__).'/../Missing';
		$app = new ConsoleRunnerStub(array('yarrow', $classpath, 'MyDocs'));
		$this->assertExitFailure($app->runApplication());		
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader(), $output);
		$this->assertContains('File not found', $output);
	}

	function testEmptyArgumentsMessage() {
		$app = new ConsoleRunnerStub(array('yarrow'));
		$this->assertExitSuccess($app->runApplication());
		$output = ob_get_contents();
		$this->assertStringStartsWith($this->getVersionHeader() , $output);
		$config = Configuration::instance();
		$this->assertContains($_ENV['PWD'], $config->inputTargets);
		$this->assertEquals($_ENV['PWD'].'/docs', $config->outputTarget);
	}
}

File: test/console_runner_test.rb
gem "minitest"
require "minitest/autorun"
require "stringio"
require "yarrow"

class ConsoleRunnerTest < Minitest::Test  
  
  SUCCESS = 0
  FAILURE = 1
  
  def test_version_message_short
    output_buffer = StringIO.new
    app = Yarrow::ConsoleRunner.new ['yarrow', '-v'], output_buffer
    assert_equal SUCCESS, app.run_application
    assert_equal "Yarrow 0.1.0\n", output_buffer.string
  end
  
  def test_version_message_long
    output_buffer = StringIO.new
    app = Yarrow::ConsoleRunner.new ['yarrow', '--version'], output_buffer
    assert_equal SUCCESS, app.run_application
    assert_equal "Yarrow 0.1.0\n", output_buffer.string
  end

  def test_help_message_short
    output_buffer = StringIO.new
    app = Yarrow::ConsoleRunner.new ['yarrow', '-h'], output_buffer
    assert_equal SUCCESS, app.run_application
    assert_includes output_buffer.string, "Yarrow 0.1.0\n"
    assert_includes output_buffer.string, "Path to the generated documentation"
  end
  
  def test_help_message_long
    output_buffer = StringIO.new
    app = Yarrow::ConsoleRunner.new ['yarrow', '--help'], output_buffer
    assert_equal SUCCESS, app.run_application
    assert_includes output_buffer.string, "Yarrow 0.1.0\n"
    assert_includes output_buffer.string, "Path to the generated documentation"
  end

  def test_invalid_short_option_message
    output_buffer = StringIO.new
    app = Yarrow::ConsoleRunner.new ['yarrow', '-j'], output_buffer
    assert_equal FAILURE, app.run_application
    assert_includes output_buffer.string, "Unrecognized option: -j"
  end
  
  def test_invalid_long_option_message
    output_buffer = StringIO.new
    app = Yarrow::ConsoleRunner.new ['yarrow', '--invalid'], output_buffer
    assert_equal FAILURE, app.run_application
    assert_includes output_buffer.string, "Unrecognized option: --invalid"
  end

  def test_invalid_long_option_message
    output_buffer = StringIO.new
    app = Yarrow::ConsoleRunner.new ['yarrow', '--invalid'], output_buffer
    assert_equal FAILURE, app.run_application
    assert_includes output_buffer.string, "Unrecognized option: --invalid"
  end
	
	def test_config_loads_if_file_exists
	  output_buffer = StringIO.new
	  test_config = File.dirname(__FILE__) + '/fixtures/test.yml'
	  
    app = Yarrow::ConsoleRunner.new ['yarrow', '--config=' + test_config], output_buffer
    assert_equal SUCCESS, app.run_application
    assert_equal app.config.meta.title, "Test Project"
	end
	
	def test_config_fails_if_file_is_missing
	  output_buffer = StringIO.new
	  test_config = File.dirname(__FILE__) + '/fixtures/missing.yml'
	  
    app = Yarrow::ConsoleRunner.new ['yarrow', '--config=' + test_config], output_buffer
    assert_equal FAILURE, app.run_application
    assert_includes output_buffer.string, "No such file or directory"
	end
  
end